### PR TITLE
CSRF in Plone

### PR DESCRIPTION
**Pull request by jensens**

### CODE REVIEW

**Comment by jensens** [comment]
This is valid to some degree for Volto based sites and RestAPI endpoints too. I am not sure if this would be better placed in the backend section. opinions?

File: docs/classic-ui/csrf.md
Comment (stevepiercy) [inline docs/classic-ui/csrf.md]:
```suggestion
In Plone, CSRF protection is done almost transparently by [`plone.protect`](https://pypi.org/project/plone.protect/).
```

File: docs/classic-ui/csrf.md
Comment (stevepiercy) [inline docs/classic-ui/csrf.md]:
```suggestion
When a logged-in user requests a page, Plone automatically includes the CSRF token in all forms by applying a transform (using `plone.transformchain`) that adds a hidden input with its value set to the token.
```

File: docs/classic-ui/csrf.md
Comment (stevepiercy) [inline docs/classic-ui/csrf.md]:
```suggestion
- control panels
```

File: docs/classic-ui/csrf.md
Comment (stevepiercy) [inline docs/classic-ui/csrf.md]:
```suggestion
- custom z3c forms
```

File: docs/classic-ui/csrf.md
Comment (stevepiercy) [inline docs/classic-ui/csrf.md]:
```suggestion
To ensure that code that is not part of a database transaction—such as code that writes to an external API or a service that is not automatically included in the transaction mechanism—is protected, you will need to manually implement protection for that code.
```

File: docs/classic-ui/csrf.md
Comment (stevepiercy) [inline docs/classic-ui/csrf.md]: Dedent all of the above. The first line of code should always be flush left.

File: docs/classic-ui/csrf.md
Comment (stevepiercy) [inline docs/classic-ui/csrf.md]: `...` is not valid Python and the lexer will choke on it. We might need to replace it with `pass` to make it valid.

```suggestion
```

File: docs/classic-ui/csrf.md
Comment (stevepiercy) [inline docs/classic-ui/csrf.md]:
```suggestion
Checks whether the request is an HTTP POST request, and raises `Unauthorized` if not.
```

File: docs/classic-ui/csrf.md
Comment (stevepiercy) [inline docs/classic-ui/csrf.md:78]: See above. It might need `pass`.

```suggestion
```

File: docs/classic-ui/csrf.md
Comment (stevepiercy) [inline docs/classic-ui/csrf.md]: We use sentence case for headings now, per `make vale`.

```suggestion
## How to add a CSRF-token to a link or form
```

File: docs/classic-ui/csrf.md
Comment (stevepiercy) [inline docs/classic-ui/csrf.md]:
```suggestion
To add a hidden field with a token to a form in a template, the above view can be used as follows:
```

File: docs/classic-ui/csrf.md
Comment (stevepiercy) [inline docs/classic-ui/csrf.md]:
```suggestion
2. If an attribute of the object is a "persistent" attribute (for example, a `PersistentDict` or `PersistentList` instance, a `BTree`, or an `annotation`), use this instead.
```

File: docs/classic-ui/csrf.md
Comment (stevepiercy) [inline docs/classic-ui/csrf.md]:
````suggestion
```{note}
This is the preferred method for allowing modification and writing of specific objects to the database.
```
````

File: docs/classic-ui/csrf.md
Comment (stevepiercy) [inline docs/classic-ui/csrf.md]:
```suggestion
    "keywords": "CSRF, security, token, protection, Cross-Site Request Forgery"
```

File: docs/classic-ui/csrf.md
Comment (stevepiercy) [inline docs/classic-ui/csrf.md:14]: Let's add the following to the Glossary as well. I anticipate we will need to refer to it elsewhere.


```
CSRF
Cross-Site Request Forgery
    Cross-Site Request Forgery (CSRF or XSRF) is a type of web attack that allows an attacker to send malicious requests to a web application on behalf of a legitimate user.
    The attack works by tricking the user's web browser into sending a request to the web application that the user did not intentionally make.
    This can allow an attacker to perform actions on the web application without the user's knowledge or consent.
    In Plone, CSRF protection is done almost transparently by [`plone.protect`](https://pypi.org/project/plone.protect/).
```

**Review by stevepiercy — CHANGES_REQUESTED** [review]
A few style, English and MyST syntax and grammar suggestions.

File: docs/classic-ui/csrf.md
Comment (jensens) [reply docs/classic-ui/csrf.md]: Either the ellipsis is needed or a pass, otherwise its not valid Python.

**Comment by stevepiercy** [comment]
> This is valid to some degree for Volto based sites and RestAPI endpoints too. I am not sure if this would be better placed in the backend section. opinions?

Good thought, and I agree, with a huge "however". In general, stuff under Classic UI or Volto is specific to the user interface, and not backend stuff. However, how does Volto implement CSRF protection with its myriad forms? I never thought about this until now. Yikes, that could be a security nightmare!

Let's look at https://demo.plone.org/login. I do not see any CSRF protection in forms. I just peed a little. 😱

@sneridagh @tiberiuichim can you provide insight to how Volto protects against CSRF?

File: docs/classic-ui/csrf.md
Comment (jensens) [reply docs/classic-ui/csrf.md]: https://realpython.com/python-ellipsis/#in-short-use-the-ellipsis-as-a-placeholder-in-python

File: docs/classic-ui/csrf.md
Comment (jensens) [reply docs/classic-ui/csrf.md:14]: I added an entry. Question now is, does it provide any value to references it in the csrf.md?

File: docs/classic-ui/csrf.md
Comment (stevepiercy) [reply docs/classic-ui/csrf.md]: We should use `pass`. [Ellipsis](https://docs.python.org/3/library/constants.html#Ellipsis) has a different meaning in Python.

Comment (stevepiercy) [reply docs/classic-ui/csrf.md]: Actually, I take that back. `...` is fine. I read that article on RealPython, and now it make sense.

File: docs/classic-ui/csrf.md
Comment (jensens) [reply docs/classic-ui/csrf.md:78]: same as above

File: docs/classic-ui/csrf.md
Comment (stevepiercy) [reply docs/classic-ui/csrf.md:14]: Not really, because it is already explicitly defined in the text, which is good. I anticipate it will be useful elsewhere in the future.

File: docs/classic-ui/csrf.md
Comment (stevepiercy) [reply docs/classic-ui/csrf.md:78]: This is fine.

**Review by stevepiercy — DISMISSED** [review]
This is excellent. Thank you!

We need to wait to hear back from the @plone/volto-team about how it protects against CSRF, to determine whether this belongs in Backend or Classic UI. Once we have that response, we can move it around if needed, and merge it.

Thank you!

File: docs/classic-ui/csrf.md
Comment (mauritsvanrees) [inline docs/classic-ui/csrf.md]:
```suggestion
from plone.protect import protect
```

File: docs/classic-ui/csrf.md
Comment (mauritsvanrees) [inline docs/classic-ui/csrf.md]: `token` needs to be defined for the next line.

```suggestion
<tal:authenticator tal:define="token context/@@authenticator/token">
```

**Review by mauritsvanrees — CHANGES_REQUESTED** [review]
Mostly LGTM. Two small fixes with inline suggestions.

**Review by mauritsvanrees — DISMISSED** [review]

File: docs/classic-ui/csrf.md
Comment (stevepiercy) [inline docs/classic-ui/csrf.md]:
```suggestion
## How to add a CSRF token to a link or form
```

File: docs/classic-ui/csrf.md
Comment (stevepiercy) [inline docs/classic-ui/csrf.md]:
```suggestion
To pass a CSRF token you need either to:
```

File: docs/classic-ui/csrf.md
Comment (stevepiercy) [inline docs/classic-ui/csrf.md]: Blackify

```suggestion
token_query = urlencode({"_authenticator": createToken()})
```

File: docs/classic-ui/csrf.md
Comment (stevepiercy) [inline docs/classic-ui/csrf.md]:
```suggestion
new_query = f"{parsed_url.query}&{token_query}"
```

**Review by stevepiercy — CHANGES_REQUESTED** [review]
Just a couple of quick fixes, which I think I can merge through GitHub's UI, and we're good to go. Thank you!

**Review by stevepiercy — APPROVED** [review]
Yup, I was able to do the quick fixes through the UI. After the build passes, I will merge.

**Comment by stevepiercy** [comment]
I decided that waiting for a response from @plone/volto-team would needlessly hold up this PR, and it is better to get it out there now. I created a new issue in their repo to give it more visibility. Volto team members seldom respond in this repo. `¯\_(ツ)_/¯`

**Pull request merged**
